### PR TITLE
[FLINK-9143] [client] Restart strategy defined in flink-conf.yaml is ignored

### PR DESCRIPTION
**Pull request by yuqi1129**
## What is the purpose of the change

This change is to fix the bug that if restart strategy was not specifically set in user code, flink will not use default strategy set in flink-conf.yaml in flip6 mode. 
## Brief change log

  - In `ClusterClient.getJobGraph`, if no strategy was set in user code, we set strategy strategy just as flink-conf.yaml do 

## Verifying this change

This change added tests and can be verified as follows:

  - Add `ClusterClientTest.testgetJobGraph` 
 -  Add `FailureRateRestartStrategyTest.testFailureRateRestartStrategyConf` and `FixedDelayRestartStrategyTest.testFixedRestartStrategyConf`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by yuqi1129** [comment]
@tillrohrmann Can you help to look if this change is OK ? thanks

**Comment by dawidwys** [comment]
@yuqi1129 Thank you for your contribution. 

Did I understand this change correctly that it only affects situation when the checkpointing is disabled? Currently in this case the `RestartStrategy` will be set from cluster's `flink-conf.yaml`. After your change the cluster configuration will never be taken into account (neither with nor without checkpointing). Therefore I would be against merging it in current shape, as it does not address the original problem, but just changes only one of many corner cases.

What do you think about such strategy for resolving `RestartStrategy`?:

1. job config
  * set in code
  * client configuration
  * (no default, even in case of checkpointing)
2. cluster config
  * cluster configuration (flink-conf)
  * default value = FixedDelayRestart (might be only in case of checkpointing)

**Comment by dawidwys** [comment]
One comment first. Please reply to review comments in github, otherwise it is hard to follow the discussion. Comments from github are posted to the jira issue automatically.

I still think your changes only affect the situation when checkpointing is disabled. If checkpointing is enabled the default strategy is set on the client side via: `ClusteClient#getJobGraph > StreamGraph#getJobGraph > StreamingJobGraphGenerator#createJobGraph > StreamingJobGraphGenerator#configureCheckpointing.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 1]
```diff
@@ -250,6 +259,35 @@ public void start() throws Exception {
...
+	 *
+	 * @param job  job graph
+	 * @param configuration configuration in flink-conf.yaml
+	 * @param isClusterPoint whether this is client side or cluster site
+	 */
+	public static void setJobgraphRestartStrategy(JobGraph job, Configuration configuration, boolean isClusterPoint) {
+
+		try {
+			ExecutionConfig config = job.getSerializedExecutionConfig().deserializeValue(
+				Dispatcher.class.getClassLoader());
```
Comment (dawidwys) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: You should use userclassloader here  (especially in case of cluster side)

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 2]
```diff
@@ -896,6 +898,10 @@ public static JobGraph getJobGraph(Configuration flinkConfig, FlinkPlan optPlan,
 			job = gen.compileJobGraph((OptimizedPlan) optPlan);
 		}
 
+		//  if we disable checkpoint and do not set restart strategy, Restart strategy will be set as in flink-conf.yaml
+		//  in flip6, jobmaster do not set this conf, so we have set this conf here.
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: before flip6, cluster site will set restart strategy:
```
    val restartStrategy =
          Option(jobGraph.getSerializedExecutionConfig()
            .deserializeValue(userCodeLoader)
            .getRestartStrategy())
            .map(RestartStrategyFactory.createRestartStrategy)
            .filter(p => p != null) match {
            case Some(strategy) => strategy
            case None => restartStrategyFactory.createRestartStrategy()
          }
```
in Jobmanger.java

in flip6, cluster will directly use jobgraph deserialized from client size

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 3]
```diff
@@ -255,6 +264,35 @@ public void start() throws Exception {
 		}
 	}
 
+	/**
+	 *
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 4]
```diff
@@ -255,6 +264,35 @@ public void start() throws Exception {
...
+	 * @param job  job graph
+	 * @param configuration configuration in flink-conf.yaml
+	 * @param isClusterPoint whether this is client side or cluster site
+	 */
+	public static void setJobgraphRestartStrategy(JobGraph job, Configuration configuration, boolean isClusterPoint) {
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: en. Indeed this is not a proper place

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: I am pretty sure the behaviour is still there. In JobMaster.java, or am I wrong?

	final RestartStrategies.RestartStrategyConfiguration restartStrategyConfiguration =
			jobGraph.getSerializedExecutionConfig()
					.deserializeValue(userCodeLoader)
					.getRestartStrategy();

	this.restartStrategy = (restartStrategyConfiguration != null) ?
			RestartStrategyFactory.createRestartStrategy(restartStrategyConfiguration) :
			jobManagerSharedServices.getRestartStrategyFactory().createRestartStrategy();

[the remark this answers is not in the document]
Comment (yuqi1129) [reply flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: You are right, I omit this point, sorry

**Comment by dawidwys** [comment]
Hi @yuqi1129 , do you still want to work on this?

**Comment by yuqi1129** [comment]
@dawidwys 
Sorry , I will push my code in two days,  if there still have any problem, any suggestions are appreciated. 
any way, if you have try to fix this problem, just go head.

**Comment by dawidwys** [comment]
Fixed in 57872d53c4584faace6dc8e4038ad1f2d068a453
